### ISSUE TITLE
imdb returning incomplete person pages

### ISSUE THREAD

**Issue by meeotch** [cinemagoer/cinemagoer#418]
#### Issue description

At some point recently, I noticed that a script I wrote using Cinemagoer was intermittently failing with keyErrors on items that I got using get_person().  I've written a little chunk of code that demonstrates the problem.

#### Version of Cinemagoer, Python and OS

- **Python:** Python 3.6.9
- **Cinemagoer:** 2022.11.22
- **OS:** ('Linux', '5.4.0-131-generic', '#147~18.04.1-Ubuntu SMP Sat Oct 15 13:10:18 UTC 2022', 'x86_64', 'x86_64')

#### Steps to reproduce the issue

Run the following code.  Typically, it will fail to find 'imdbID' in the item returned from get_person() the first time.  It sleeps a few seconds, then tries again.  

```python
#!/usr/bin/env python

from imdb import Cinemagoer
import time

ia = Cinemagoer()
personid = '0006904'
item = None
while not item:
    item = ia.get_person(personid)
    print(item.keys())
    if 'imdbID' not in item:
        print("Trying person %s again..." % personid)
        item = None
        time.sleep(3)
```

Sample output - Note how the item keys are a subset of the full keys at first, then eventually become the full set:

```
['headshot', 'birth name', 'height', 'mini biography', 'trivia', 'quotes', 'salary history', 'birth date', 'birth notes', 'full-size headshot']
Trying person 0006904 again...
['birth info', 'headshot', 'known for', 'filmography', 'imdbID', 'name', 'imdbIndex', 'birth name', 'height', 'mini biography', 'trivia', 'quotes', 'salary history', 'birth date', 'birth notes', 'canonical name', 'long imdb name', 'long imdb canonical name', 'full-size headshot']
```
#### What's the expected result?

I expect the person to have the full set of keys.

#### What's the actual result?

You have to re-try get_person() a few times to get the full set of keys.

#### Additonal Information:
I'm not sure if this is related, but I also noticed that using requests.get() to retrieve a "normal" imdb page (non-reference - i.e. one that doesn't end in "/reference") has also started failing for me intermittently, as imdb has added a Captcha.  The fix was to insert a time.sleep(5) before requests.get().  So I guess imdb has been throttling requests that come too quickly.  Maybe it's also serving up "light" versions of pages to Cinemagoer, if too many requests are made too quickly?

**Comment by alberanid** [comment]
thanks for the report, closing as duplicate of https://github.com/cinemagoer/cinemagoer/issues/415

I've just committed a potential fix, please test again using the repository version.

**Comment by meeotch** [comment]
Reposting this, since I accidentally posted it to the wrong issue the first time:

I know that there are now several open issues on related topics, but one more data point: As of today, or the past couple days, it's gotten worse. My test script above now fails consistently many times in a row.

I've also noticed that tinyMediaManger comes up with no results when scraping with the imdb scraper. A week ago, I'd be able to re-submit, and get results. But now I have to leave it idle for 5-10min in order to get results.

So it does feel like some sort of spam protection / throttling on the part of imdb.com.

UPDATE:  my test script above now seems to fail completely on every attempt.  I did have one thought about a fix, if it is in fact some sort of anti-scraping attempts by imdb.com:  Is it possible that logging in to an imdb.com account would work around it?  I'm not sure if that's easy or even possible without pulling in a bunch of nonsense like Selenium, though.  Maybe manually logging in via a browser, then saving out the cookies and loading them in Cinemagoer.  Though possibly that too would be defeated by the SessionID.  Just thinking out loud, since Cinemagoer is such a useful tool.

**Comment by alberanid** [comment]
@meeotch currently all the tests in the test suite complete succesfully.

Please double check that you are using the latest Cinemagoer version from this repository.

Also, consider that Cinemagoer is a tool to fetch a single resource at a time, and is not thought to allow mass-scraping.

So, if you hit some limits, I guess the only option is to greatly reduce the rate at which you request information.

Regarding the ability to login: it would be nice to have, but it will be even more easier to block it completely.

There is partial support for cookies, that was used a long time ago to support the ability to search for adult titles.

Basically, you have to look at your browser resources once logged in and look for the "id" and "uu" cookies and pass them as `cookie_id` and `cookie_uu` args of `imdb.Cinemagoer()`; however, it seems that now the "id" cookie no longer exists, so I do not know if it's still possible to emulate a login just based on cookies. (but that would be a matter for a different issue)

**Comment by meeotch** [comment]
I can confirm that the test script, and my larger project both seem to be working now.  Don't even need the pauses anymore.  I did need to update my rule for the "Known For" section of the person page, which I added to movieparser.  (I'm using it to grab the few top titles that imdb thinks are most interesting for a person, so that I don't have to list the full filmography in the pages my project outputs.)  Will paste the code below, in case it's useful to anyone.

Interesting tidbit:  on the TinyMediaManager reddit, the dev of that project mentioned that imdb is returning two different types of search pages now, plus the other various format changes you've already been working on.  So maybe the user-agent and the format changes were the problem all along, rather than some more sophisticated attempt by imdb to block scraping.

`        Rule(
            key='known for',
            extractor=Rules(
                foreach='//div[@data-testid="nm_flmg_kwn_for"]//a[@class="ipc-primary-image-list-card__title"]',
                rules=[
                    Rule(
                        key='link',
                        extractor=Path('./@href')
                    ),
                    Rule(
                        key='title',
                        extractor=Path('./text()')
                    )
                ],
                transform=lambda x: build_movie(
                    x.get('title') or '',
                    movieID=analyze_imdbid(x.get('link') or '')
                )
            )
        ),
`
